### DISCUSSION TITLE
Database Seeder with Relationship

### DISCUSSION THREAD

**Discussion by liam-edwards** [adonisjs/core#2703] [Help]
I have two Models; `User` and `Role`. `User` has a `Role` and it must exist. I have seeders for both models but how do I set the role within the seeder?

Below is what I would have expected it to be.
```ts
await User.updateOrCreateMany('email', [
    name: 'Liam',
    email: 'myemail@example.com',
    password: 'A password',
    role: Role.find(1),
])
```
Or even `role_id: 1`, but neither work.

How can I define the related column within the seeder?

**Comment by thetutlage** [comment]
I recommend reading the docs to see how you can persist relationships. https://docs.adonisjs.com/guides/models/relationships#create-relationships

**Reply by liam-edwards** [reply]
Forgive me if I'm missing something but I don't see anything there that helps. I have read through the docs and there doesn't appear to be anything on how to define a relationship in a seeder using `updateOrCreateMany`.

**Comment by aayush123** [comment]
Is the role seeder running before user seeder?
If yes, you should be able to do `roleId: 1` in your updateOrCreate call.

In my case, I had a many to many relationship between User & Role.
So I created the role, then the user using firstOrCreate and then just attached role to user after creation

**Reply by liam-edwards** [reply]
Except `roleId: 1` doesn't work. Everything I try leads to the same error message:

```
$ node ace db:seed
❯ error     database\seeders\MainSeeder\Index
  Value for the "id" is null or undefined inside "updateOrCreateMany" payload
```

**Reply by aayush123** [reply]
What relationship do you have between role & user?

**Reply by liam-edwards** [reply]
HasOne. Here are the two models.

```ts
import { DateTime } from 'luxon'
import Hash from '@ioc:Adonis/Core/Hash'
import {
	column,
	beforeSave,
	BaseModel,
	hasOne,
} from '@ioc:Adonis/Lucid/Orm'
import Role from 'App/Models/Role'
import { HasOne } from '@ioc:Adonis/Lucid/Relations'

export default class User extends BaseModel {
	@column({ isPrimary: true })
	public id: number

	@column()
	public name: string

	@column()
	public email: string

	@column({ serializeAs: null })
	public password: string

	@column()
	public rememberMeToken?: string

	@hasOne(() => Role)
	public role: HasOne<typeof Role>

	@column.dateTime({ autoCreate: true })
	public createdAt: DateTime

	@column.dateTime({ autoCreate: true, autoUpdate: true })
	public updatedAt: DateTime

	@beforeSave()
	public static async hashPassword(user: User) {
		if (user.$dirty.password) {
			user.password = await Hash.make(user.password)
		}
	}
}
```

```ts
import { DateTime } from 'luxon'
import { BaseModel, column } from '@ioc:Adonis/Lucid/Orm'

export default class Role extends BaseModel {
	@column({ isPrimary: true })
	public id: number

	@column()
	public name: string

	@column()
	public accessLevel: number

	@column.dateTime({ autoCreate: true })
	public createdAt: DateTime

	@column.dateTime({ autoCreate: true, autoUpdate: true })
	public updatedAt: DateTime
}
```

**Reply by aayush123** [reply]
You can use [set-related](https://docs.adonisjs.com/reference/orm/relations/has-one#setrelated) to set role for user

**Reply by liam-edwards** [reply]
What? How does that work with `updateOrCreateMany`? The docs need to be _a lot_ more clear in how this works because I'm confused as all hell.

**Reply by aayush123** [reply]
First, in which model are you keeping the foreign key? User or Role?
I'll provide an explanation after this

**Reply by liam-edwards** [reply]
The `users` table has a `role_id` column referencing `roles.id`.

**Reply by aayush123** [reply]
So for starters, you should have a belongsTo relationship in User Model.
I'm assuming a role can be associated to many users, so you should create a hasMany relationship to users in Role Model.
Further, you should try persisting users using relationship persistence on Roles after creating the above relationships.
As clearly mentioned in the docs, `role.related('users').updateOrCreateMany('email', [...])` should do the job after above changes.
Note that role here is what you get after `await Role.find(1)`

**Reply by liam-edwards** [reply]
Now I'm getting that initial error again... Here are my seeders, hopefully you can help me out.

```ts
import BaseSeeder from '@ioc:Adonis/Lucid/Seeder'
import Application from '@ioc:Adonis/Core/Application'

export default class IndexSeeder extends BaseSeeder {
	private async runSeeder(seeder: { default: typeof BaseSeeder }) {
		/**
		 * Do not run when not in dev mode and seeder is development only
		 */
		if (seeder.default.developmentOnly && !Application.inDev) {
			return
		}

		await new seeder.default(this.client).run()
	}

	public async run() {
		await this.runSeeder(await import('../RoleSeeder'))
		await this.runSeeder(await import('../UserSeeder'))
	}
}
```

```ts
import BaseSeeder from '@ioc:Adonis/Lucid/Seeder'
import Role from 'App/Models/Role'

export default class RoleSeeder extends BaseSeeder {
	public async run() {
		await Role.updateOrCreateMany('id', [
			{
				name: 'super-admin',
				accessLevel: 100,
			},
			{
				name: 'admin',
				accessLevel: 90,
			},
			{
				name: 'editor',
				accessLevel: 80,
			},
		])
	}
}

```

```ts
import BaseSeeder from '@ioc:Adonis/Lucid/Seeder'
import Role from 'App/Models/Role'

export default class UserSeeder extends BaseSeeder {
	public async run() {
		const role = await Role.findOrFail(1)
		role.related('users').updateOrCreateMany([
			{
				name: 'Liam',
				email: 'myemail@example.com',
				password: 'A password',
			},
		], 'email')
	}
}
```

**Reply by aayush123** [reply]
Add await to the `role.related...` call.
If it still doesn't work, I'd advise creating a reproducible github repo. Going back and forth with code snippets gets cumbersome after a while

**Reply by liam-edwards** [reply]
Here, https://github.com/liam-edwards/adonis-seeder-bug.

Just need to set your environment variables, run the migration, and witness the error on `node ace db:seed`

**Reply by aayush123** [reply]
So there are various things wrong in your code, as listed below:

- You're missing the `roleId` column in User Model. **The relationship will not work without this column**
- You're using incorrect column name `accessLevel` in your migration, it should be `access_level`, unless you specify  a CamelCase [Naming Strategy](https://docs.adonisjs.com/reference/orm/naming-strategy#sidenav-open)
- Most importantly, your RoleSeeder was using an incorrect clause in `updateOrCreateMany` call.

As mentioned in the docs, the first argument in the `updateOrCreateMany` call is the search criteria.
You are passing `'id'` into it, which means you wish to search using the _id_ and update the ones that match, and create the ones that don't.

However, **you are not passing id key in the payload**. How would you expect the ORM to search for records to update or create?
You should either pass `'name'` or `'access_level'` in the search criteria.

Updating this successfully runs the RoleSeeder.

Furthermore, the bug you were facing was apparently in the _RoleSeeder itself_, not the UserSeeder (a little glance at the error logs tells you that).

Afterwards, the solutions mentioned in the thread work and create Users as expected.

I have done all these changes in the [PR ](https://github.com/liam-edwards/adonis-seeder-bug/pull/1).
Have a look at the diff and understand the corrections, read the docs and try to understand the methods you are using, the structure and the logic behind the code you are using. It'll help you out in the long run.

**Reply by liam-edwards** [reply]
Yup, I saw and replied to the PR already. The case error of `access_level` was definitely a mistake that I would have noticed had it gotten that far. As for missing `roleId` in the User model, coming from Laravel, I was used to not needing that so that was a complete oversight.

My main error, which you already pointed out, was using the wrong predicate in the RoleSeeder's `updateOrCreateMany`. When I saw the error message, I thought it was referring to me not specifying an id value within the UserSeeder.

I did read the docs, but I think a couple of things just escaped me or I happened to skim past a line or two mentioning it.

Thanks again for all your help.

**Reply by aayush123** [reply]
No worries. Happy hacking! 👍

**Comment by thetutlage** [comment]
I think docs are super clear and covers the broad API. The API for seeders is no different than using a model somewhere else.

First thing, the following code cannot work for several reasons

```ts
await User.updateOrCreateMany('email', [
    name: 'Liam',
    email: 'myemail@example.com',
    password: 'A password',
    role: Role.find(1),
])
```

- `Role.find` returns a promise and hence you need to await it.
- `Role.find` if row is found returns a model instance. If you are trying to supply the model instance directly, then for sure you haven't read the docs.

Last, your relationship says. "User -> HasOne -> Role". But the role model doesn't have a `userId` column on it. I am not sure how you can expect the relationship to work, if you don't define the foreignKeys.

Finally, this all boils down to understanding the basics of how relationships work and reading docs vs criticizing docs
